### PR TITLE
Refuse to generate elements that conflict with some "important" obj-c…

### PR DESCRIPTION
**Pull request by chamons**
… selectors

- https://github.com/mono/Embeddinator-4000/issues/346
- hash and class so far

### CODE REVIEW

File: objcgen/objcgenerator-postprocessor.cs [hunk 1]
```diff
@@ -60,10 +69,25 @@ protected IEnumerable<ProcessedProperty> PostProcessProperties (IEnumerable<Prop
...
 			}
 		}
 
+		bool ProcessPotentialName (ProcessedProperty processedProperty)
+		{
+			switch (processedProperty.Name) {
+				case "setClass":
+				case "class":
+					Delayed.Add (ErrorHelper.CreateWarning (1051, $"Element '{processedProperty.Property.Name}' is not generated because its name conflicts with important objective-c selector '{processedProperty.Name}'"));
+					return false;
```
Comment (rolfbjarne) [inline objcgen/objcgenerator-postprocessor.cs]: I think you can add all the selectors in `NSObjectProtocol`: https://developer.apple.com/reference/objectivec/1418956-nsobject?language=objc

File: objcgen/objcgenerator-postprocessor.cs [hunk 2]
```diff
@@ -35,20 +35,29 @@ protected IEnumerable<ProcessedMethod> PostProcessMethods (IEnumerable<Processed
...
 			if (IsOperatorOrFriendlyVersion (method)) {
 				string nameOverride = OperatorOverloads.GetObjCName (processedMethod.Method.Name, processedMethod.Method.ParameterCount);
 				if (nameOverride != null)
 					processedMethod.NameOverride = nameOverride;
 			}
+
+			string objCSignature = processedMethod.GetObjcSignature ();
+			switch (objCSignature) {
+			case "hash":
+				Delayed.Add (ErrorHelper.CreateWarning (1051, $"Element '{processedMethod.Method.Name}' is not generated because its name conflicts with important objective-c selector '{objCSignature}'"));
```
Comment (rolfbjarne) [inline objcgen/objcgenerator-postprocessor.cs]: Properties can be named `Hash` as well.

**Comment by chamons** [comment]
I added all of NSObjectProtocol and improved the lookup to hopefully catch more cases. I need to add more tests cases, but before that the question needs to be - Do we skip generation or rename?

As @spouliot pointed out, without renaming support, it is difficult to call one of these without code changes. On the other hand, it can get rather confusing if you have hash and _hash for example.

**Comment by chamons** [comment]
build

**Comment by rolfbjarne** [comment]
I don't think there's any way we can rename these selectors to something that's actually nice, so I'd go with something descriptive.

What about `managed_X`?

**Comment by spouliot** [comment]
Fwiw stuff like 'hash' should not confuse objects developers. The rename is to avoid confusing the objc runtime :) 
Still a descriptive name can't do harm

**Comment by chamons** [comment]
I decided to see how difficult it would be to implement attribute based renames:

https://github.com/mono/Embeddinator-4000/pull/366

Let me add a rename to managed_Foo for this PR in any case.

**Comment by chamons** [comment]
So underscore is a bit tricky due to this - https://github.com/mono/Embeddinator-4000/blob/objc/objcgen/processedtypes.cs#L235 since properties _are_ special types. 

I'll just do managedFoo for now.

**Comment by spouliot** [comment]
What would happen if a managedFoo exists ?

File: docs/errors.md [hunk 3]
```diff
@@ -209,6 +209,14 @@ There should be an earlier warning giving more information why type `T` is not s
 
 Note: Supported features will evolve with new versions of the tool.
 
+<h3><a name="EM1051"/>Element `E` is generated instead as `F` because its name conflicts with an important objective-c selector.</h3>
+
+This is a **warning** that the element `E` will be generated instead as `F` because its name conflicts with an important objective-c selector.
+
+Selectors such as "hash" and "class" have important meaning in objective-c and must be overridden carefully.
```
Comment (spouliot) [inline docs/errors.md]: Let's document the NSObject protocol

File: docs/errors.md [hunk 4]
```diff
@@ -209,6 +209,14 @@ There should be an earlier warning giving more information why type `T` is not s
...
+
+Note: Supported features will evolve with new versions of the tool.
+
```
Comment (spouliot) [inline docs/errors.md]: Supported features -> the list of reserved selectors ...

**Comment by chamons** [comment]
> What would happen if a managedFoo exists ?

We very likely generate garbage code. :(

That is until https://github.com/mono/Embeddinator-4000/issues/369 get handled.

**Comment by chamons** [comment]
@spouliot Because I was not handling fields, which are different that properties apparently. :(

I'll have to rework it to include fields.

**Pull request merged**